### PR TITLE
Decouple "Active Terminal" and "Focused Control"

### LINKED ISSUE #522: Active terminal does not get focus when window gets focus
**Your Windows build number:**
10.0.18362.86

**What you're doing and what's happening:**

1. Open Windows Terminal
2. Start PowerShell console.
2. Make sure that the window have lost focus
3. Click on the window to focus the application
4. Start writing
5. Input is not sent to the active console

**What's wrong / what should be happening instead:**
Input should be sent to the active console.

### LINKED ISSUE #999: If you click on the separators between panes, then no panes will be focused
Self explanatory. This is related to #528, but is definitely different.

How does this interact with #994? If we make the separator a margin, then this won't be an issue.

### LINKED ISSUE #1205: Untie "active control" from "currently XAML-focused control"
> I think there are some issues in tying "active control" to "control currently focused in the XAML stack".. what do you think? Should we just holistically fix this all up by making active/focused not the same?

A tab should have a notion of its "active control", even if that control isn't currently focused. Right now, we have some problematic mixing of user intent and UI state that cause us to have tabs with no active control and tabs whose titles flicker when they become "focused" even though what's "active" hasn't changed.

Changes in which control is active can be propagated out through a single "ActiveTerminalChanged" event.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Unties the concept of "focused control" from "active control".

Previously, we were exclusively using the "Focused" state of `TermControl`s to determine which one was active. This was fraught with gotchas - if anything else became focused, then suddenly there was _no_ pane focused in the Tab. This happened especially frequently if the user clicked on a tab to focus the window. Furthermore, in experimental branches with more UI added to the Terminal (such as [dev/migrie/f/2046-command-palette](https://github.com/microsoft/terminal/tree/dev/migrie/f/2046-command-palette)), when these UIs were added to the Terminal, they'd take focus, which again meant that there was no focused pane.

This fixes these issue by having each Tab manually track which Pane is active in that tab. The Tab is now the arbiter of who in the tree is "active". Panes still track this state, for them to be able to MoveFocus appropriately. 

It also contains a related fix to prevent the tab separator from stealing focus from the TermControl. This required us to set the color of the un-focused Pane border to some color other that Transparent, so I went with the TabViewBackground. Panes now look like the following:

image

## References

See also: #2046

## PR Checklist
* [x] Closes #1205
* [x] Closes #522
* [x] Closes #999
* [x] I work here
* [😢] Tests added/passed
* [n/a] Requires documentation to be updated

## Validation Steps Performed

Tested manually opening panes, closing panes, clicking around panes, the whole dance.

### CODE REVIEW

File: src/cascadia/TerminalApp/Tab.h [hunk 1]
```diff
@@ -36,15 +35,22 @@ class Tab
...
 
     bool _focused{ false };
     winrt::Microsoft::UI::Xaml::Controls::TabViewItem _tabViewItem{ nullptr };
 
+    std::function<void()> _pfnActivePaneChanged{ nullptr };
+
     void _MakeTabViewItem();
     void _Focus();
+
+    void _AttachEventHandersToControl(const winrt::Microsoft::Terminal::TerminalControl::TermControl& control);
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Tab.h]:
```suggestion
    void _AttachEventHandlersToControl(const winrt::Microsoft::Terminal::TerminalControl::TermControl& control);
```

File: src/cascadia/TerminalApp/Pane.cpp [hunk 2]
```diff
@@ -1053,4 +1065,79 @@ Size Pane::_GetMinSize() const
...
+// - Register a callback function to be called when this Pane's control gains
+//   focused. This is used by the Tab hosting us so it can track which Pane in
+//   the tree was the last one to be focused.
+// Arguments:
+// - pfnGotFocus: A function that should be called when this pane's control gets focus.
+// Return Value:
+// - <none>
+void Pane::SetGotFocusCallback(std::function<void(std::shared_ptr<Pane>)> pfnGotFocus)
+{
+    _pfnGotFocus = pfnGotFocus;
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/Pane.cpp]:
```suggestion
    _pfnGotFocus = std::move(pfnGotFocus);
```

Comment (DHowett-MSFT) [reply src/cascadia/TerminalApp/Pane.cpp]: you're taking a copy in as a param, you can elide another copy by moving here.

File: src/cascadia/TerminalApp/Pane.h [hunk 3]
```diff
@@ -43,57 +43,74 @@ class Pane : public std::enable_shared_from_this<Pane>
...
 
     bool CanSplit(SplitState splitType);
-    void Split(SplitState splitType, const GUID& profile, const winrt::Microsoft::Terminal::TerminalControl::TermControl& control);
+    void Split(SplitState splitType,
+               const GUID& profile,
+               const winrt::Microsoft::Terminal::TerminalControl::TermControl& control);
 
     void Close();
 
+    void SetGotFocusCallback(std::function<void(std::shared_ptr<Pane>)> pfnGotFocus);
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/Pane.h]: Hey, if we make this a `winrt::event`, we get lightweight delegates _and_ an easy 1:1 change when we eventually make Pane a winrt class.

Comment (DHowett-MSFT) [reply src/cascadia/TerminalApp/Pane.h]: (I believe we can 100% use `winrt::event` without being ourselves a winrt class)

File: src/cascadia/TerminalApp/Tab.cpp [hunk 4]
```diff
@@ -276,8 +284,42 @@ void Tab::NavigateFocus(const winrt::TerminalApp::Direction& direction)
...
+// - control: the TermControl to add events to.
+// Return Value:
+// - <none>
+void Tab::_AttachEventHandersToControl(const TermControl& control)
+{
+    control.TitleChanged([this](auto newTitle) {
+        // The title of the control changed, but not necessarily the title
+        // of the tab. Get the title of the active pane of the tab, and set
+        // the tab's text to the active panes' text.
+        auto newTabTitle = GetFocusedTitle();
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/Tab.cpp]: should GetFocused become GetActive?

File: src/cascadia/TerminalApp/Pane.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.cpp]: Should I do the same with `Tab::SetActivePaneChangedCallback`?

File: src/cascadia/TerminalApp/Pane.h (hunk 3 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.h]: Uhg. I tried so hard to get that to work before finding that the type of the event needs to be `winrt::event<winrt::delegate<std::shared_ptr<Pane>>>` ;__;

I'll fix this up, this is much better.

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.h]: I take it back, this is _not_ much better.

Right now I'm propagating the callback down to the leaves, then back up to the parent when leaves are closed.

With a `winrt::event`, I'm having a much harder time getting this right. Having the Tab register a `GotFocus` handler on the panes works just fine, but when the panes are closed, they just _will not_ activate the right pane in the tree. I'll keep playing with it.

**Comment by zadjii-msft** [comment]
@DHowett-MSFT

1. I suppose, back down the rabbit hole I go
2. Oh yea that looks wrong. The whole lambda in `Split` should be gone, that should just be the `AttachPaneEventHandlers(second)` call

**Pull request merged**